syft/frameworks/torch/tensors/interpreters/replicated_shared.py

```diff
@@ -122,12 +112,120 @@ def private_linear_operation(self, secret, operator):
+        }
+        shares_map = self.get_shares_map()
+        for player in players:
+            shares_map[player] = (
+                operator(shares_map[player][0], plain_text_map[player]),
+                operator(shares_map[player][1], plain_text_map[player]),
+            )
+        return ReplicatedSharingTensor(shares_map)
+
+    def private_multiplication_operation(self, secret, operator):
+        x, y = self.get_shares_map(), secret.get_shares_map()
+        players = self.get_players()
+        z = [
+            operator(x[player][0], y[player][0])
+            + operator(x[player][1], y[player][0])
+            + operator(x[player][0], y[player][1])
+            for player in players
+        ]
+        z = self.__add_noise(z)
+        z = self.__reshare(z, players)
+        return ReplicatedSharingTensor(z)
+
+    @staticmethod
+    def __add_noise(shares):
+        noisy_shares = [share + gen_alpha_3of3(share.location).wrap() for share in shares]
+        return noisy_shares
+
+    @staticmethod
+    def __reshare(shares, workers):
+        """convert 3-out-of-3 secret sharing: {player i : share i}
+          to 2-out-of-3 sharing: {player i : (share i, share i+1)}  """
+        shares_map = {}
+        for i in range(len(shares)):
+            pointer = shares[(i + 1) % len(shares)].copy().move(workers[i])
+            shares_map[workers[i]] = (shares[i], pointer)
+        return shares_map
+
+    def conv2d(self, value, padding=0):
+        batches_in, channels_in, width_in, height_in = value.shape
+        batches_out, channels_out, width_out, height_out = self.shape
+        value = self.unfold(value, height_out, padding)
+        self = self.apply_to_shares(torch.Tensor.view, channels_out, -1)
+        result = self @ value
+        size = (height_in - height_out + 2 * padding) + 1
+        result = result.apply_to_shares(torch.Tensor.view, batches_out, channels_out, size, size)
+        return result
+
+    def unfold(self, value, kernel_size, padding=0):
+        return self.__switch_public_private(
+            value, self.public_unfold, self.private_unfold, kernel_size, padding
+        )
+
+    @staticmethod
+    def public_unfold(plain_text, kernel_size, padding):
+        plain_text = plain_text.double()
+        plain_text = torch.nn.functional.unfold(
+            plain_text, kernel_size=kernel_size, padding=padding
+        )
+        plain_text = plain_text.long()
+        return plain_text
+
+    @staticmethod
+    def private_unfold(secret, kernel_size, padding):
+        secret = secret.apply_to_shares(torch.Tensor.double)
+        secret = secret.apply_to_shares(
+            torch.nn.functional.unfold, kernel_size=kernel_size, padding=padding
+        )
+        secret = secret.apply_to_shares(torch.Tensor.long)
+        return secret
+
+    @staticmethod
+    def __switch_public_private(value, public_function, private_function, *args, **kwargs):
```

Comment:
I really like this function :D

Reply:
yeah me too :D it saved a lot of code repetition